sources/models/src/modeled_types/ecs.rs

```diff
@@ -0,0 +1,172 @@
+use lazy_static::lazy_static;
+use regex::Regex;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+// Just need serde's Error in scope to get its trait methods
+use serde::de::Error as _;
+use snafu::ensure;
+use std::borrow::Borrow;
+use std::convert::TryFrom;
+use std::fmt;
+use std::ops::Deref;
+use super::error;
+
+/// ECSAttributeKey represents a string that contains a valid ECS attribute key.  It stores
+/// the original string and makes it accessible through standard traits.
+// https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_Attribute.html
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct ECSAttributeKey {
+    inner: String,
+}
+
+
+// The name of the attribute. Up to 128 letters (uppercase and lowercase), numbers, hyphens,
+// underscores, and periods are allowed.
+lazy_static! {
+    pub(crate) static ref ECS_ATTRIBUTE_KEY: Regex = Regex::new(
+        r"(?x)^
+       [a-zA-Z0-9-_\./]{1,128}
+   $"
+    )
+    .unwrap();
+}
+
+impl TryFrom<&str> for ECSAttributeKey {
+    type Error = error::Error;
+
+    fn try_from(input: &str) -> Result<Self, Self::Error> {
+        ensure!(
+            ECS_ATTRIBUTE_KEY.is_match(input),
+            error::BigPattern {
+                thing: "ECS attribute key",
+                input
+            }
+        );
+        Ok(ECSAttributeKey {
+            inner: input.to_string(),
+        })
+    }
+}
+
+string_impls_for!(ECSAttributeKey, "ECSAttributeKey");
+
+#[cfg(test)]
+mod test_ecs_attribute_key {
+    use super::ECSAttributeKey;
+    use std::convert::TryFrom;
+
+    #[test]
+    fn good_keys() {
+        for ok in &[
+            "a",
+            "alphabetical",
+            "1234567890",
+            "with-dash",
+            "have.period/slash",
+            "have_underscore_too",
+            &"a".repeat(128),
+            ".leadingperiod",
+            "trailingperiod.",
+        ] {
+            ECSAttributeKey::try_from(*ok).unwrap();
+        }
+    }
+
+    #[test]
+    fn bad_keys() {
+        for err in &[
```

Comment:
`err` is a some what misleading name for this.
```suggestion
        for key in &[
```